fabric-entity-tick-events-v1/src/main/java/net/fabricmc/fabric/impl/event/entity/TickEventInternals.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.impl.event.entity;
+
+import java.util.IdentityHashMap;
+import java.util.Map;
+
+import com.google.common.base.Preconditions;
+
+import net.minecraft.entity.Entity;
+import net.minecraft.util.profiler.Profiler;
+
+import net.fabricmc.fabric.api.event.Event;
+import net.fabricmc.fabric.api.event.EventFactory;
+import net.fabricmc.fabric.api.event.v1.EntityTickCallback;
+
+public class TickEventInternals {
+	/** Cache of entity class to {@code Event}. */
+	private static final Map<Class<?>, CascadingEvent> ENTITY_TICK_EVENTS = new IdentityHashMap<>();
+
+	/**
+	 * Retrieves an entity tick event for a given class. If none exists, this method creates one and registers it
+	 * as a cascading child of the superclass' event.
+	 *
+	 * <p>The returned {@code Event} will have new callbacks registered automatically whenever one is registered
+	 * to a parent event.
+	 *
+	 * @return a tick {@code Event} for entities of the given class
+	 */
+	@SuppressWarnings("unchecked")
+	public static <E extends Entity> CascadingEvent<EntityTickCallback<E>> getOrCreateEntityEvent(Class<?> entityClass) {
+		Preconditions.checkArgument(Entity.class.isAssignableFrom(entityClass), "Cannot register entity tick callback to non-entity class. Found: " + entityClass.getName());
```

Comment:
Hmm, what if a mod lants to add callback to interfaces like `Hoglin` `VillagerDataContainer` etc to perform common logic?

Reply:
Those special cases would add tons of complexity since we only accept classes that explicitly extend `Entity`. If you wanted to target that, I would say use `MobEntityWithAi` and instanceof check with `Hoglin` or `VillagerDataContainer`.

Reply:
sure, that's fine.